File: src/leetcode/python/easy/PathSum_112.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class PathSum_112(object):
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        def path_sum(node, tmp, sum):
            if not node: return False
            tmp += node.val
            if not node.left and not node.right:
                return True if tmp == sum else False
            if path_sum(node.left, tmp, sum): return True
            return path_sum(node.right, tmp, sum)

        tmp = 0
        return path_sum(root, 0, sum)

File: src/leetcode/java/PathSum_112.java
package leetcode.java;

public class PathSum_112 {
  public boolean hasPathSum(TreeNode root, int sum) {
    return pathSum(root, 0, sum);
  }

  public static boolean pathSum(TreeNode root, int tmp, int sum) {
    if (root == null) return false;
    tmp += root.val;
    if (root.left == null && root.right == null) {
      if (tmp == sum) return true;
      else return false;
    }
    if (pathSum(root.left, tmp, sum)) return true;
    return pathSum(root.right, tmp, sum);
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
